File: src/classes/Human.java
package classes;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Created by pranavkonduru on 9/4/18.
 */
public class Human implements LifeForm, Mammals {

    private String nameCapitalized;

    public Human(@NotNull String name) {
        System.out.println("Hello " +name);
    }

    public Human(@NotNull String name, @NotNull String gender, @Nullable Integer age){
        this(name);
        if (age != null) {
            System.out.println("Age given is: "+age);
        }

        nameCapitalized = name.toUpperCase();
    }

    @NotNull
    @Override
    public Boolean move() {
        return true;
    }

    @NotNull
    @Override
    public String speak() {
        return MammalsDefaultImpls.speak();
    }

    @Nullable
    @Override
    public Integer limbs() {
        return 2;
    }

    public String getCapitalizedName() throws Exception {
        if (nameCapitalized != null) {
            return nameCapitalized;
        }

        throw new Exception("Not initialized yet");
    }
}

File: src/classes/Human.kt
package classes

/**
 * Created by pranavkonduru on 9/4/18.
 */
class Human(val name: String) : LifeForm, Babies {

    /**
     * Initialization block for the primary constructor
     */
    init {
        println("Name supplied is $name")
    }

    /**
     * Secondary constructor with nullability check
     */
    constructor(name: String, gender: String, age: Int?): this(name){
        age?.also { println("Age given is $it")}
    }

    /**
     * Normal override
     */
    override fun move(): Boolean {
        return true
    }

    /**
     * Overriding method from interface
     * that has a default code block
     *
     * Calling super is optional
     */
    override fun limbs(): Int? {
        return 2
    }

    /**
     * If two interfaces have methods with same signatures and names,
     *
     * return type should specify which interface it is calling.
     */
    override fun speak(): String {
        return super<Babies>.speak()
    }
}